### PR TITLE
Add controller code for attenuator.

### LINKED ISSUE #9: Add attenuator controller
Refactor and open source the attenuator controller code.

### PR DESCRIPTION
**Pull request by xpconanfan**
* Add an attenuator device lib implemented for minicircuit device.
* The minicircuit implementation showcases the interface expected
  for an attenuator controller.
* Move keys only used by the attenuator module out of mobly.keys.

fixes #9

### CODE REVIEW

File: mobly/controllers/attenuator.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    def __init__(self, attn_device, idx=0, name=None):
+        self.model = attn_device.model
+        self.attn_device = attn_device
+        self.idx = idx
+        if (self.idx >= attn_device.path_count):
+            raise IndexError(
+                "Attenuator index out of range for attenuator attn_device")
+
+    def set_attn(self, value):
```
Comment (dthkao) [inline mobly/controllers/attenuator.py]: Same for these. Why shorten the method names?

File: mobly/controllers/attenuator.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        """This function sets the attenuation of Attenuator.
+
+        Args:
+            value: This is a floating point value for nominal attenuation to be
```
Comment (dthkao) [inline mobly/controllers/attenuator.py]: units? db?

File: mobly/controllers/attenuator.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import importlib
+import logging
+
+MOBLY_CONTROLLER_CONFIG_NAME = "Attenuator"
+KEY_ADDRESS = "Address"
```
Comment (dthkao) [inline mobly/controllers/attenuator.py]: This assumes any attenuator device will be telnet. Is there any chance one might use a different connection method?

Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: No, this is just a macro.
When a new way of connection is needed, can just add more macros

File: mobly/controllers/attenuator.py (hunk 1 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: 'attn' and 'atten' have been mixed among doc, code, and reference.
I'm making the abbreviated reference consistent everywhere by using 'attn'

File: mobly/controllers/attenuator_lib/minicircuits.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+from mobly.controllers.attenuator_lib import telnet_client
+
+
+class AttenuatorDevice(object):
+    """This provides a specific telnet-controlled implementation of
+    AttenuatorDevice for Mini-Circuits RC-DAT attenuators.
+
+    Attributes:
+        path_count: The number of signal attenuation path this device has.
+        _telnet_client: The telnet client used to communicate with the
```
Comment (adorokhine) [inline mobly/controllers/attenuator_lib/minicircuits.py]: I don't think private attributes are typically documented in the user-facing module documentation. If this comment is valuable, it should probably be a comment at the variable declaration site.

File: mobly/controllers/attenuator_lib/telnet_client.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Helper module for common telnet capability to communicate with
+AttenuatorDevice(s).
+
+User code shouldn't need to directly access this class.
+"""
+
+import telnetlib
```
Comment (adorokhine) [inline mobly/controllers/attenuator_lib/telnet_client.py]: Does this require setup.py changes?

File: mobly/controllers/attenuator_lib/telnet_client.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+from mobly.controllers import attenuator
+
+
+def _ascii_string(uc_string):
+    return str(uc_string).encode('ASCII')
+
+
+class TelnetClient(object):
+    """This is an internal helper class for Telnet+SCPI command-based
```
Comment (adorokhine) [inline mobly/controllers/attenuator_lib/telnet_client.py]: Is there anything specific to SCPI in this class? If so should it be telnet_scpi_client?

File: mobly/controllers/attenuator.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+KEY_PORT = "Port"
+
+
+def create(configs):
+    objs = []
+    for c in configs:
```
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: Could you avoid abbreviating "c" here? Makes the long loop a little hard to follow.

File: mobly/controllers/attenuator.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        attn_model = c["Model"]
```
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: android_device has lowercase keys right? Could we be consistent?

File: mobly/controllers/attenuator.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        # Default to telnet.
+        protocol = c.get("Protocol", "telnet")
+        # Import the correct driver module for the attenuator device
```
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: Not sure why we need to do this import; won't register_controller(minicircuits) take care of this?

File: mobly/controllers/attenuator.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        module_name = "mobly.controllers.attenuator_lib.%s" % attn_model
+        module = importlib.import_module(module_name)
+        # Create each
+        path_cnt = c["PathCount"]
+        attn_device = module.AttenuatorDevice(path_cnt)
+        attn_device.model = attn_model
+        insts = attn_device.open(c[KEY_ADDRESS], c[KEY_PORT])
```
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: instances

File: mobly/controllers/attenuator.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    objs = []
+    for c in configs:
+        attn_model = c["Model"]
+        # Default to telnet.
+        protocol = c.get("Protocol", "telnet")
+        # Import the correct driver module for the attenuator device
+        module_name = "mobly.controllers.attenuator_lib.%s" % attn_model
+        module = importlib.import_module(module_name)
+        # Create each
+        path_cnt = c["PathCount"]
```
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: What is a PathCount? A single attenuator can have more than one path?

File: mobly/controllers/attenuator.py (hunk 3 above)
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: You have a lot of inline keys in the module. I'm ok with that (no need to make them all constants) but could you add an example config fragment showing what it should look like?

Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Done

File: mobly/controllers/attenuator.py (hunk 7 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Done

File: mobly/controllers/attenuator.py (hunk 8 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: which case do we want to choose?

File: mobly/controllers/attenuator.py (hunk 9 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: No, you would not register minicircuits directly, you would register this module, which in turn loads the appropriate 'driver' based on the config

File: mobly/controllers/attenuator.py (hunk 10 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Done

File: mobly/controllers/attenuator_lib/minicircuits.py (hunk 4 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator_lib/minicircuits.py]: Done

File: mobly/controllers/attenuator.py (hunk 2 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Done

File: mobly/controllers/attenuator.py (hunk 9 above)
Comment (adorokhine) [inline mobly/controllers/attenuator.py]: What are the advantages of implementing another system of indirection versus just using register_controller?

File: mobly/controllers/attenuator.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        self.attenuators[0].set_attn(0, 50)
+        self.attenuators[0].set_attn(1, 40)
+
+    The benefit the former is that the physical test bed can use either four
+    single-channel attenuators, or one four-channel attenuators. Whereas the
+    latter forces the test bed to use a four-channel attenuator.
+    """
+    def __init__(self, attn_device, idx=0, name=None):
+        self.model = attn_device.model
+        self.attn_device = attn_device
```
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Spelling out every 'attenuation' is gonna make things unnecessarily long?
should be ok as long as it's consistent?
and set_attn and get_attn seem pretty straightforward?

Comment (dthkao) [inline mobly/controllers/attenuator.py]: The varnames you chose in the example code, the classnames, and package names spell it out, and I think it makes it clearer. Are we really worried about verbosity?

File: mobly/controllers/attenuator.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    for config in configs:
+        attn_model = config["Model"]
+        # Import the correct driver module for the attenuator device
+        module_name = "mobly.controllers.attenuator_lib.%s" % attn_model
+        module = importlib.import_module(module_name)
+        # Create each
+        path_cnt = config["PathCount"]
+        attn_device = module.AttenuatorDevice(path_cnt)
+        attn_device.model = attn_model
+        instances = attn_device.open(config[KEY_ADDRESS], config[KEY_PORT])
```
Comment (dthkao) [inline mobly/controllers/attenuator.py]: This is what I mean as assuming the attenuator device will be telnet.

Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Yes it is.
We can easily change this later if we need to support a different driver. And no test code needs to change if someone chooses to try running with a different hardware.
The point is to avoid test scripts having to change because hardware changed.

Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: But if you want to abstract it out, we can do that later.

File: mobly/controllers/attenuator.py (hunk 12 above)
Comment (xpconanfan) [inline mobly/controllers/attenuator.py]: Discussed offline, use 'atten' instead of 'attn' because 'attn' can also mean other things like 'attention'
